### ISSUE TITLE
provide `autofs` role

### ISSUE THREAD

**Issue by riccardomurri** [gc3-uzh-ch/elasticluster#405]
Write an `autofs` role that installs `autofs5` and configures it to be able to
auto-mount exported shares from arbitrary hosts on the network under the `/net`
directory hierarchy.

In other words, the deployed configuration should:

- include a map `/etc/auto.net` into `/etc/auto.master`;
- `/etc/auto.net` should allow for subdirectories `/net/nfs3`, `/net/nfs4` etc that mount according to protocol;
- then accessing e.g. `/net/nfs4/master001/home` mounts `master001:/home` via NFSv4

For reference, this is the same setup we have on Hydra and the other nodes of
the local compute cluster. One challenge might be to have the same config run
fine across Debian/Ubuntu and CentOS 6+7.

**Comment by riccardomurri** [comment]
For reference, this is the `/etc/auto.master` to drive both NFSv3 and NFSv4 auto-mounts:

```
# /etc/auto.master
#
# Master map for autofs(8); it is used to define mountpoints
# managed by `autofs`, and the corresponding map file. When
# changed, a HUP signal (`service autofs reload`) must be
# sent to the `autofs` daemon. See `man 5 auto.master` for details.
#

/net/nfs3  -hosts
/net/nfs4  program:/etc/auto.nfs4  hard,intr,nosuid,nodev

/site      file:/etc/auto.site     --timeout=1200
```

**Comment by riccardomurri** [comment]
This is file `/etc/auto.nfs4` referenced above. **Note** that it must have
permissions `a+rx` otherwise it will be taken as a map/configuration file, not
as a _script outputting the right map item_.

```
#!/bin/sh

# Debian's `autofs` is not able to mount NFSv4 shares, as it does not
# strip the root filesystem prefix.  For more information, see:
#
# - https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=646625
# - https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=533893

# add "nosymlink" here if you want to suppress symlinking local filesystems
# add "nonstrict" to make it OK for some filesystems to not mount
# other NFSv4 mount options can be specified as options to the map in auto.master
opts="-fstype=nfs4"

echo "${opts} ${1}:/"
```

Also note that CentOS 6+7 might require a different script to work.

**Comment by riccardomurri** [comment]
This is script `/etc/auto.smb` with a similar purpose (completely untested)::

```
#!/bin/bash

# This file must be executable to work! chmod 755!

# Automagically mount CIFS shares in the network, similar to
# what autofs -hosts does for NFS.

# Put a line like the following in /etc/auto.master:
# /cifs  /etc/auto.smb --timeout=300
# You'll be able to access Windows and Samba shares in your network
# under /cifs/host.domain/share

# "smbclient -L" is used to obtain a list of shares from the given host.
# In some environments, this requires valid credentials.

# This script knows 2 methods to obtain credentials:
# 1) if a credentials file (see mount.cifs(8)) is present
#    under /etc/creds/$key, use it.
# 2) Otherwise, try to find a usable kerberos credentials cache
#    for the uid of the user that was first to trigger the mount
#    and use that.
# If both methods fail, the script will try to obtain the list
# of shares anonymously.

get_krb5_cache() {
    cache=
    uid=${UID}
    for x in $(ls -d /run/user/$uid/krb5cc_* 2>/dev/null); do
        if [ -d "$x" ] && klist -s DIR:"$x"; then
	    cache=DIR:$x
            return
        fi
    done
    if [ -f /tmp/krb5cc_$uid ] && klist -s /tmp/krb5cc_$uid; then
	    cache=/tmp/krb5cc_$uid
	    return
    fi
}

key="$1"
opts="-fstype=cifs"

for P in /bin /sbin /usr/bin /usr/sbin
do
	if [ -x $P/smbclient ]
	then
		SMBCLIENT=$P/smbclient
		break
	fi
done

[ -x $SMBCLIENT ] || exit 1

creds=/etc/creds/$key
if [ -f "$creds" ]; then
    opts="$opts"',uid=$UID,gid=$GID,credentials='"$creds"
    smbopts="-A $creds"
else
    get_krb5_cache
    if [ -n "$cache" ]; then
        opts="$opts"',multiuser,cruid=$UID,sec=krb5i'
        smbopts="-k"
        export KRB5CCNAME=$cache
    else
        opts="$opts"',guest'
        smbopts="-N"
    fi
fi

$SMBCLIENT $smbopts -gL "$key" 2>/dev/null| awk -v "key=$key" -v "opts=$opts" -F '|' -- '
	BEGIN	{ ORS=""; first=1 }
	/Disk/	{
		  if (first)
			print opts; first=0
		  dir = $2
		  loc = $2
		  # Enclose mount dir and location in quotes
		  # Double quote "$" in location as it is special
		  gsub(/\$$/, "\\$", loc);
		  gsub(/\&/,"\\\\&",loc)
		  print " \\\n\t \"/" dir "\"", "\"://" key "/" loc "\""
		}
	END 	{ if (!first) print "\n"; else exit 1 }
	'
```

**Comment by riccardomurri** [comment]
ATM we only need NFS to work; CIFS/SMB can wait.

**Comment by riccardomurri** [comment]
This has been implemented in commit c3f73f13010e8f851fc95a9c6fb474f48034fe22

Current status:

* automouting NFS works
* home directories on clusters are currently *not* auto-mounted
* support for auto-mouting GCS using [`gcsfuse`]()https://github.com/GoogleCloudPlatform/gcsfuse is there but does not work due to [a bug](https://github.com/GoogleCloudPlatform/gcsfuse/issues/309) in `gcsfuse`
* I'll open a separate issue for adding SMB/CIFS support
